### PR TITLE
Tacho structure symmetric LU solver for CPU and GPU

### PR DESCRIPTION
**Pull request by kyungjoo-kim**
## Motivation

Two things are considered for this refactoring: 
- AMD build
  - Adding HIP enabling macro and determine the code path for HIP
- LU factorization
  - This is work-in-progress.
- Test without tasking infrastructure
  - On GPUs, we do not want to use async tasking but the infrastructure is still required 
  - AMD does not have the infrastructure. For further future work, we first remove the dependency on tasking.

## Related Issues

* Is blocked by 
  #10383 - Trilinos common package gtest does not work when CUDA is used. This can triggers to fail the PR test.
  #10366 - Kokkos OpenMP tasking is not activated when CUDA RDC is not ON. For now, I have work-around and this is not urgent.

## Stakeholder Feedback

@crdohrm For the incremental development and testing purpose, you can use this branch or wait until it is merged.

## Testing

The code relies on Trilinos gtest package and it does not have use correct the cuda SM flag. This may triggers to fail the PR test as well. For my own test, I override the flag and it passes the unit tests.

### CODE REVIEW

**Review by crdohrm — APPROVED** [review]
Let me know when everything is ready to go.

**Comment by kyungjoo-kim** [comment]
It passes all tests except for unknown failure [Krino_krino_unit_MPI_4](https://trilinos-cdash.sandia.gov/test/2779444). 
[cuda test output](https://trilinos-cdash.sandia.gov/index.php?project=Trilinos&filtercount=2&showfilters=1&filtercombine=and&field1=buildname&compare1=63&value1=10406&field2=buildstamp&compare2=63&value2=Pull%20Request)

Could someone from @trilinos/framework explain this failure ?

**Comment by kyungjoo-kim** [comment]
@trilinos/framework cuda PR test does not run. Is there any system issue ?

**Comment by ZUUL42** [comment]
@kyungjoo-kim, [it is running](https://trilinos-cdash.sandia.gov/index.php?project=Trilinos&display=project&filtercount=2&showfilters=1&filtercombine=and&field1=buildname&compare1=63&value1=_cuda-10.1.243&field2=buildstamp&compare2=63&value2=Pull%20Request). There is a known issue with the Krino_krino_unit_MPI_4 test failing, we'll be discussing that at stand up today, but I'm [showing a lot more tests failing in the build for this PR](https://trilinos-cdash.sandia.gov/index.php?project=Trilinos&display=project&filtercount=3&showfilters=1&filtercombine=and&field1=site&compare1=61&value1=vortex52&field2=buildname&compare2=61&value2=PR-10406-test-ats2_cuda-10.1.243-gnu-8.3.1-spmpi-rolling_release_static_Volta70_Power9_no-asan_no-complex_no-fpic_mpi_pt_no-rdc_uvm_deprecated-on_no-package-enables-714&field3=buildstamp&compare3=61&value3=20220406-0616-Pull%20Request).

**Comment by kyungjoo-kim** [comment]
@ZUUL42 if you take a look at the details of the failed tests, it is all about the launch failure of the tests.
```
BEFORE: jsrun  '-p' '4' '--rs_per_socket' '4' '/vscratch1/trilinos/jenkins/vortex-trilinos/workspace/trilinos-folder/_cuda_10.1.243@2/pull_request_test/packages/amesos/example/Amesos_example_AmesosFactory_HB.exe'
AFTER: export TPETRA_ASSUME_CUDA_AWARE_MPI=1; jsrun  '-E LD_PRELOAD=/usr/tce/packages/spectrum-mpi/ibm/spectrum-mpi-rolling-release/lib/pami_451/libpami.so' '-M -gpu' '-p' '4' '--rs_per_socket' '4' '/vscratch1/trilinos/jenkins/vortex-trilinos/workspace/trilinos-folder/_cuda_10.1.243@2/pull_request_test/packages/amesos/example/Amesos_example_AmesosFactory_HB.exe'
Warning: jsrun wrapper does not handle unexpected option '-E LD_PRELOAD=/usr/tce/packages/spectrum-mpi/ibm/spectrum-mpi-rolling-release/lib/pami_451/libpami.so', may prevent LLNL-specific error detection
         If option '-E LD_PRELOAD=/usr/tce/packages/spectrum-mpi/ibm/spectrum-mpi-rolling-release/lib/pami_451/libpami.so' valid,  please tell John Gyllenhaal (gyllen@llnl.gov, 4-5485)
 
ERROR: ld.so: object '/usr/tce/packages/spectrum-mpi/ibm/spectrum-mpi-rolling-release/lib/pami_451/libpami.so' from LD_PRELOAD cannot be preloaded: ignored.
ERROR: ld.so: object '/usr/tce/packages/spectrum-mpi/ibm/spectrum-mpi-rolling-release/lib/pami_451/libpami.so' from LD_PRELOAD cannot be preloaded: ignored.
ls: cannot access /var/tmp/jsm.vortex59.91724/2621517/JSM_rm_port_91724_2621517: No such file or directory
Error: CSM reports jsmd running on host vortex59, but can't locate  there.
04-06-2022 04:37:50:978 51161 main: Error initializing RM connection. Exiting.
jsrun return value: 1
```

**Comment by ZUUL42** [comment]
@kyungjoo-kim I follow now.
Interesting. Both the 3000+ test failure and the single test failure involve the same
`ERROR: ld.so: object '/usr/tce/packages/spectrum-mpi/ibm/spectrum-mpi-rolling-release/lib/pami_451/libpami.so' from LD_PRELOAD cannot be preloaded: ignored.`
We'll have to look into this. I believe someone is looking into the krino test failure, but I'll have to see how the rest of this relates.

**Comment by kyungjoo-kim** [comment]
The Krino test failure is same. jsrun fails to launch the test.

**Comment by kyungjoo-kim** [comment]
@ZUUL42 I will no longer put "retest". I have no luck and the system failure on cuda is consistent. Please let me know if the PR tester gets back to be normal.

**Comment by kyungjoo-kim** [comment]
@ZUUL42 Can I ping about the status of PR tester ? Is there any tracking issue about failed test of cuda, etc ?

**Comment by e10harvey** [comment]
> @ZUUL42 I will no longer put "retest". I have no luck and the system failure on cuda is consistent. Please let me know if the PR tester gets back to be normal.

@kyungjoo-kim: Once #10429 merges, the krino failure will be resolved. If you're seeing other mass test failures, please report them to the cuda machine team or let us know so we can report them.

**Comment by ZUUL42** [comment]
I think hit an issue we had with ascic158 filling up.
That has been cleared up since the failing tests started.
Next run should be clear.

**Comment by kyungjoo-kim** [comment]
@trilinos/zoltan2 @trilinos/zoltan @ZUUL42 

https://testing.sandia.gov/cdash/viewTest.php?onlyfailed&buildid=10910481

There are repeated Zoltan test failures that are not related to this PR.

**Comment by kyungjoo-kim** [comment]
In particular, when I looked at the details of the test output, I don't understand the test logic. It failed for one test and overall it says the final test is passed. 
```
Test 7944:partkway-v3 FAILED (Diff failed on 4 files)
Test 7944:  0 out of 1 tests PASSED.
Test 7944:  1 out of 1 tests FAILED.

--------------------------------------------------------------------------------

TEST_0: Return code = 0
TEST_0: Pass criteria = Zero return code [PASSED]
TEST_0: Result = PASSED

================================================================================

OVERALL FINAL RESULT: TEST PASSED (Zoltan_ch_7944_parmetis_parallel)

```

**Comment by ZUUL42** [comment]
@kyungjoo-kim, I've been seeing those tests fail in other builds. We're looking into it.

**Comment by kyungjoo-kim** [comment]
the failed error is compiler error from intrepid2. not related to this.

**Comment by kyungjoo-kim** [comment]
@ZUUL42 @e10harvey @trilinos/framework The PR test keeps failing with system failures. From the series of resets, 
- the last PR test does not start coda tests and python-3 tests.
- intel and clang are failed to start build
- gcc debug build failed due to time out 
- gcc 8.3 and 7.2 serial build pass the tests.
- [zoltan test failure due to time out](https://testing.sandia.gov/cdash/viewTest.php?onlyfailed&buildid=11089647) is not related to this PR. 

I am not sure if I should keep trying retest. This needs to be merged to Trillions for ASC. @crdohrm

**Comment by e10harvey** [comment]
@kyungjoo-kim: This failure is due to https://github.com/trilinos/Trilinos/pull/10598 and has been resolved. Please try retesting now.

CC: @trilinos/framework

**Comment by e10harvey** [comment]
@kyungjoo-kim: The latest failures are related to https://github.com/trilinos/Trilinos/pull/10598. I'm looking into it.

**Comment by kyungjoo-kim** [comment]
@e10harvey I cannot submit a Trilinos ticket. Whenever I create a ticket, it returns me to search website for help. Anyway, it keeps failing with external conditions.

**Comment by e10harvey** [comment]
> @e10harvey I cannot submit a Trilinos ticket. Whenever I create a ticket, it returns me to search website for help. Anyway, it keeps failing with external conditions.

@ccober6, @wadeburgess.

**Comment by kyungjoo-kim** [comment]
```
In terms of the test failures, I see that a couple transient errors were encountered (network failures during git clone). Please let the CI tests run again after resolving these build and test failures reported to CDash: https://trilinos-cdash.sandia.gov/index.php?project=Trilinos&parentid=169828&filtercount=1&showfilters=1&field1=buildstarttime&compare1=84&value1=NOW&filtercombine=and.
```

@trilinos/framework @e10harvey The build failure is from Zoltan and it is not related to this PR at all.

**Comment by kyungjoo-kim** [comment]
@ZUUL42 @e10harvey @trilinos/framework In the previous PR test, it fails to build owing to a missing TPL header.

```
/scratch/trilinos/jenkins/ascic141/workspace/Trilinos_PR_python3/Trilinos/packages/seacas/libraries/ioss[exodus/Ioex_BaseDatabaseIO.h:16](https://github.com/trilinos/Trilinos//blob/master/exodus/Ioex_BaseDatabaseIO.h#L16):10: fatal error: exodusII.h: No such file or directory
```

[here the information](https://trilinos-cdash.sandia.gov/viewBuildError.php?buildid=179750)

File: packages/shylu/shylu_node/tacho/unit-test/Tacho_TestDenseLinearAlgebra.hpp
Comment (kyungjoo-kim) [inline packages/shylu/shylu_node/tacho/unit-test/Tacho_TestDenseLinearAlgebra.hpp:90]: @e10harvey check this.

Comment (kyungjoo-kim) [reply packages/shylu/shylu_node/tacho/unit-test/Tacho_TestDenseLinearAlgebra.hpp:90]: now this pr passed. I just merge this first.

Comment (e10harvey) [reply packages/shylu/shylu_node/tacho/unit-test/Tacho_TestDenseLinearAlgebra.hpp:90]: @kyungjoo-kim: I understand where you're coming from. However, let's focus on getting master merge working again before introducing additional changes & features that could result in new failures that block master merge. Please move this fix into #10625.

Comment (e10harvey) [reply packages/shylu/shylu_node/tacho/unit-test/Tacho_TestDenseLinearAlgebra.hpp:90]: @kyungjoo-kim: Getting the master branch updated is a top-priority for the product leaders. I will make sure this PR is next in the priority queue once the master branch is updated.

**Review by crdohrm — APPROVED** [review]

**Comment by kyungjoo-kim** [comment]
@e10harvey Does pre-merge inspection take this much of time (about 1 hour after the pr is approved by clark) ? What does it mean by "NOT REVIEWED" in the above ?

**Comment by e10harvey** [comment]
> @e10harvey Does pre-merge inspection take this much of time (about 1 hour after the pr is approved by clark) ? What does it mean by "NOT REVIEWED" in the above ?

@kyungjoo-kim: The pre-merge inspection takes a very small amount of time but since Trilinos PR testing takes such a long time to run on other PRs we have to wait hours for the AutoTester to circle back around to your PR.

**Comment by kyungjoo-kim** [comment]
@e10harvey thanks. then, I will wait until it shows something on the pre-merge inspection check.

**Comment by kyungjoo-kim** [comment]
Finally done.

**Pull request merged**
